internals/idkey/idkey.go

```diff
@@ -0,0 +1,265 @@
+
+// Identity key must implement a crypto signer.
+var _ crypto.Signer = (*IDKey)(nil)
+
+type IDKey struct {
+	keyDir string
+	key    any
+}
+
+// New checks if an existing private identity key exists, and loads the key
+// if it does. It creates a new private identity key and persists it to disk
+// if no key was found. Cases where explicit control is desired on when to
+// generate or load can use GenerateKey and LoadKey directly.
+func New(keyDir string) (*IDKey, error) {
+	keyPath := filepath.Join(keyDir, identityKeyFile)
+	exists, err := pathExists(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot access key file %q: %w", keyPath, err)
+	}
+	if exists {
+		return LoadKey(keyDir)
+	}
+	return GenerateKey(keyDir)
+}
+
+// GenerateKey generates a new identity key and persists it to disk. This
+// function should only ever be called on the first boot otherwise the
+// existing identity will be overwritten.
+//
+// This function is equivalent to running:
+//
+//	openssl genpkey -algorithm Ed25519 -out key.pem
+func GenerateKey(keyDir string) (*IDKey, error) {
+	key := &IDKey{
+		keyDir: keyDir,
+	}
+	err := key.createDir()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create identity directory: %w", err)
+	}
+	err = key.newEd25519()
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate identity key: %w", err)
+	}
+	err = key.save()
+	if err != nil {
+		return nil, fmt.Errorf("cannot save identity key: %w", err)
+	}
+	return key, nil
+}
+
+// LoadKey loads an existing identity key from disk.
+func LoadKey(keyDir string) (*IDKey, error) {
+	key := &IDKey{
+		keyDir: keyDir,
+	}
+	// Load from disk.
+	err := key.load()
+	if err != nil {
+		return nil, fmt.Errorf("cannot load identity key: %w", err)
+	}
+	return key, nil
+}
+
+// createDir verifies the directory layout and permissions, and creates
+// the leaf element of the key driectory if it does not yet exist.
+func (k *IDKey) createDir() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return expectPermission(k.keyDir, 0o700)
+	}
+
+	// Create the leaf directory node with 0o700 permissions.
+	return os.Mkdir(k.keyDir, 0o700)
+}
+
+// Public implements the crypto.Signer.
+func (k *IDKey) Public() crypto.PublicKey {
+	signer := k.key.(crypto.Signer)
+	return signer.Public()
+}
+
+// Sign implements the crypto.Signer.
+func (k *IDKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	signer := k.key.(crypto.Signer)
+	return signer.Sign(rand, digest, opts)
+}
+
+// newEd25519 generates a new ed25519 key.
+func (k *IDKey) newEd25519() error {
+	var err error
+	_, k.key, err = ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// load loads the private identity key from storage.
+func (k *IDKey) load() error {
+	exists, err := pathExists(k.keyDir)
+	if err != nil || !exists {
+		return fmt.Errorf("directory %q is not accessible", k.keyDir)
```

Comment:
nit:
```suggestion
		return fmt.Errorf("directory %q is not accessible: %w", k.keyDir, err)
```

Reply:
I agree with @anpep suggestion here, except we probably want two different cases, because `fmt.Errorf` with `%w` yields the string `%!w(<nil>)` for nil errors. So we probably need to handle the err!=nil and the !exists cases separately.

Reply:
I now handle both cases separately. Resolving.